Hold the state

oli-obk opened the pull request:
precursor to #14

ok... iterator style was much easier than anticipated... The current design would require streaming iterators though...

Basically it returns the statement or the terminator that is going to be interpreted *next*. Then it actually interprets that at the next `step` call and returns the statement after that.

solson commented on src/interpreter/iterator.rs:
```diff
@@ -0,0 +1,92 @@
+            process: Self::dummy,
+        }
+    }
+    fn dummy(&mut self) -> EvalResult<()> { Ok(()) }
```
Style nit: blank lines between functions

solson commented:
Do you think it's problematic that the recursive parts (where `call_nested` was called before) aren't part of the iterator, but have their own iterators?

I suppose like @eddyb [said](https://github.com/solson/miri/issues/14#issuecomment-222519492) we could pre-compute statics/constants/promoted-rvalues.

Or on the other hand, in the current design once their resulting allocations are properly cached (and marked immutable in `Memory` somehow) they will effectively be lazily computed, which isn't semantically much different from pre-computing. So I guess it makes sense that they aren't part of the "same execution" (same `Stepper`).

solson commented on src/interpreter/mod.rs:
```diff
@@ -996,13 +982,13 @@ impl<'a, 'b, 'mir, 'tcx> FnEvalContext<'a, 'b, 'mir, 'tcx> {
                 match *literal {
                     Value { ref value } => Ok(self.const_to_ptr(value)?),
-                    Item { .. } => unimplemented!(),
+                    Item { .. } => Err(EvalError::Unimplemented(format!("function pointers are unimplemented"))),
```
I believe `Literal::Item` can be more than just function pointers, but I'm not sure. @eddyb?

oli-obk replied:
yea it can be more than that... but most of it is tagged with `FIXME`s

solson replied:
I'd be happy with `"literal items (e.g. function pointers) are unimplemented"` just to avoid the possibility of future confusion.

oli-obk replied:
done

oli-obk commented:
We can always add that later. Whenever we enter a function, we could add its promoted values to a Vector which we first process in the `step` function. That messes up the order somewhat, but I think it can be expected that constants are computed before everything else.

For global constants/statics we'd need something else, because that would have to happen during the creation of the global context and before the first function call... Maybe some kind of builder?

> and marked immutable in Memory somehow

not all statics should be immutable.

solson commented:
> not all statics should be immutable.

@oli-obk In the constant evaluator, according to @eddyb's model, they should be. I believe even `static mut` mutation was to be banned at compile-time.

solson commented on src/interpreter/mod.rs:
```diff
@@ -20,6 +20,8 @@ use error::{EvalError, EvalResult};
 use memory::{Memory, Pointer};
 use primval::{self, PrimVal};
 
+mod iterator;
```
This module name/filename is confusing given that it has no `Iterator` impl. `mod stepper` perhaps?

oli-obk replied:
done

solson commented on tests/compile-fail/unimplemented.rs:
```diff
@@ -0,0 +1,16 @@
+#[miri_run]
+fn static_mut() {
+    unsafe {
+        X = 6;
```
So yeah, this might fail for a different reason in the future (mutation of a static), but I'm fine with having it here so we notice when that happens.

Although it would be allowed in versions of Miri not intended for rustc const eval.

eddyb commented:
LGTM, modulo the lack of constant pre-computation. Do note that when part of this gets merged into rustc, some of the constants will already have been computed by rustc itself and may have caused compiler errors, possibly before miri can really take over.

oli-obk commented:
> some of the constants will already have been computed by rustc itself

Why only some?

> may have caused compiler errors, possibly before miri can really take over.

Interpreting code with compiler errors is probably not useful anyway. Miri [already aborts if there were compiler errors](https://github.com/solson/miri/blob/master/src/bin/miri.rs#L27).

> the lack of constant pre-computation

It's easy to implement for promoted constants. But there's no list of global constants/statics...

Also evaluating a constant might require evaluating another constant. So even if we cache constants, we'd need to evaluate them in an order that makes sure that all required constants are computed before the requiring constant.

I'm not sure how to proceed...

eddyb commented:
> Why only some?

Because you may not be able to evaluate, e.g. an associated constant used from a function which is also generic - only trans-time monomorphization (or miri, in this case) will trigger its evaluation.

> It's easy to implement for promoted constants. But there's no list of global constants/statics...

Not really, what I meant is collecting all *uses* of constants in a function, and having that as a different set of states, before any statement/terminator.
When finishing evaluation of a constant, you pop a "stack frame" and go to the next pending constant in the "caller", and if there aren't any left, you go to `(START_BLOCK, 0)`.

oli-obk commented:
I refactored the code to cache statics globally and cache promoteds just for the current stackframe. This has the unfortunate side-effect, that statics can now be written to.

I also removed the current recursion that happens when evaluating a static/constant, and now the stepper simply pushes another stack frame for constants/statics

oli-obk commented:
The way it's written now, we technically don't need to do the pre-evaluation of promoteds and constants... we could just do it on-demand.

eddyb commented:
@oli-obk How can you easily do it on-demand? You can have arbitrarily many constant operands in one statement/terminator, even the simplest ones.
This is in part because of the indexing on lvalue, i.e. `var0[A][B][C][D]` is a valid `Lvalue`.

eddyb commented:
@oli-obk Unless you mean just before the statement/terminator, which I guess is a fine approach.

oli-obk commented:
> This is in part because of the indexing on lvalue, i.e. var0[A][B][C][D] is a valid Lvalue.

oh right, so we wouldn't save the allocation of the vector... let's leave it as is.

oli-obk commented:
some benchmarks, not much of a change compared with the [last ones](https://github.com/solson/miri/pull/13#issue-157490823)

```
test fib           ... bench:         111 ns/iter (+/- 3)
test fib_iter      ... bench:           1 ns/iter (+/- 0)
test fib_iter_miri ... bench:   8,813,523 ns/iter (+/- 296,007)
test fib_miri      ... bench:   5,816,824 ns/iter (+/- 259,427)
```

oli-obk commented:
this PR ended up being able to evaluate fewer functions than before, because there might be statics/promoteds that aren't needed in the interpretation and thus weren't interpreted before. If those statics/constants contained code that miri can't interpret (yet), the entire function evaluation failed. I added a refactoring of the call stack and moved to interpret constants only just before the statement/terminator that needs them.

solson commented:
Sorry for letting this sit. I'll be reviewing it tomorrow.

solson commented on src/interpreter/stepper.rs:
```diff
@@ -0,0 +1,214 @@
+pub struct Stepper<'fncx, 'a: 'fncx, 'b: 'a + 'mir, 'mir: 'fncx, 'tcx: 'b>{
+    fncx: &'fncx mut FnEvalContext<'a, 'b, 'mir, 'tcx>,
+    mir: CachedMir<'mir, 'tcx>,
+    process: fn (&mut Stepper<'fncx, 'a, 'b, 'mir, 'tcx>) -> EvalResult<()>,
```
Can we replace the function pointers with an `enum`? Instead of calling `(self.process)(self)`, match over the enum and call whatever method is needed.

I'm not sure why we'd use a function pointer in the first place.

oli-obk replied:
> I'm not sure why we'd use a function pointer in the first place.

because ... reasons ... no idea.

I did a refactoring to get rid of the `Constant` state (I simply push all the constants' stackframes directly, instead of pushing one, evaluating it, pushing the next one, evaluating it, ...)

solson commented on src/interpreter/stepper.rs:
```diff
@@ -0,0 +1,214 @@
+
+pub struct Stepper<'fncx, 'a: 'fncx, 'b: 'a + 'mir, 'mir: 'fncx, 'tcx: 'b>{
+    fncx: &'fncx mut FnEvalContext<'a, 'b, 'mir, 'tcx>,
+    mir: CachedMir<'mir, 'tcx>,
```
Can we remove this field and just use `fncx.mir()` all the time?

oli-obk replied:
done

solson commented on src/interpreter/frame.rs:
```diff
@@ -0,0 +1,923 @@
+    pub constants: Vec<(ConstantId<'tcx>, Span, Pointer, CachedMir<'mir, 'tcx>)>,
+}
+
+pub(super) struct FrameInterpreter<'a, 'b: 'a, 'mir: 'a, 'tcx: 'a + 'mir + 'b> {
```
I don't understand the `FrameInterpreter`/`FnEvalContext` split. Why is statement evaluation here and terminator evaluation over there?

I want to lean towards fewer "context" types if possible. I wasn't really happy I had to make the `Global`/`Fn` split, either.

oli-obk replied:
I reverted it.

oli-obk replied:
And also undid the splitting into files, we can do that in a separate PR

solson commented on src/interpreter/frame.rs:
```diff
@@ -0,0 +1,923 @@
+
+    pub fn type_layout(&self, ty: Ty<'tcx>) -> &'tcx Layout {
+        self.gecx.type_layout(ty, self.frame.substs)
+    }
```
Forwarding functions like the above three are why I made `FnEvalContext` deref to the `GlobalEvalContext` (like how `TyCtxt` works in rustc).

oli-obk replied:
Yea, I tried to move as much as possible to the `GlobalEventContext`.

solson commented on src/interpreter/global.rs:
```diff
@@ -0,0 +1,298 @@
+    }
+
+    pub fn type_layout(&self, ty: Ty<'tcx>, substs: &'tcx Substs<'tcx>) -> &'tcx Layout {
+        if let Some(layout) = self.monomorphizations.borrow().get(&(ty, substs)) {
```
`monomorphizations` should cache the results of `monomorphize`, no? Sometimes we need the monomorphized type itself, and getting that should be fast, too.

Caching the `Ty` -> `Layout` steps could be done separately.

oli-obk replied:
I undid all the caching, because looking at the traces it's not really clear what's going on. Often times the "same" (textual output is the same) type + subst combo get hashed differently... I don't know enough about this do to it properly

solson replied:
@eddyb ^ Any idea what could make `Substs` hashing go wrong?

solson commented on src/interpreter/global.rs:
```diff
@@ -0,0 +1,298 @@
+            return layout;
+        }
+        debug!("monomorphising: {:?} + {:?}", ty, substs);
+        // TODO(solson): Is this inefficient? Needs investigation.
```
I think we can remove this comment now.

solson commented on src/interpreter/mod.rs:
```diff
@@ -37,17 +37,6 @@ struct GlobalEvalContext<'a, 'tcx: 'a> {
-    ///   * Function DefIds and Substs to print proper substituted function names.
-    ///   * Spans pointing to specific function calls in the source.
-    name_stack: Vec<(DefId, &'tcx Substs<'tcx>, codemap::Span)>,
-
```
Hmm, I guess cleaning up the stacks was a lot easier than I thought (particularly `substs_stack`).

We definitely need a ton more tests dealing with generic functions, monomorphization, and substs to make sure we haven't screwed it up, but your change here seems correct to me.

Thanks for the cleanups. \o/

solson commented on src/interpreter/mod.rs:
```diff
@@ -129,9 +129,9 @@ enum TerminatorTarget {
+enum ConstantId<'tcx> {
     Promoted { index: usize },
-    Static { def_id: DefId },
+    Static { def_id: DefId, substs: &'tcx Substs<'tcx> },
```
Can `Promoted` constants requires substs, too? For example:

```rust
struct Foo<T>(T);
fn foo() {
    let x = &Foo(42); // The Foo(42) constant could be promoted, right?
}
```

@eddyb ^ confirm/deny?

solson replied:
Nevermind, this is fixed in a later commit.

eddyb replied:
Why would statics though?

solson replied:
Good point. Statics can't have substs since they aren't allowed to use the function-level type parameters.

solson commented on src/interpreter/mod.rs:
```diff
@@ -178,6 +172,16 @@ impl<'a, 'b, 'mir, 'tcx> FnEvalContext<'a, 'b, 'mir, 'tcx> {
         }
     }
 
+    fn alloc_ret_ptr(&mut self, ty: ty::FnOutput<'tcx>) -> Option<Pointer> {
```
Nit: rename the `ty` argument to avoid shadowing below? Maybe `output_ty`.

oli-obk replied:
done

solson commented on src/interpreter/stepper.rs:
```diff
@@ -0,0 +1,214 @@
+            mir::Literal::Item { def_id, substs } => {
+                let item_ty = self.gecx.tcx.lookup_item_type(def_id).subst(self.gecx.tcx, substs);
+                if item_ty.ty.is_fn() {
+                    // unimplemented
```
`unimplemented!()`

oli-obk replied:
this is not necessarily executed, so if we error here, we can't evaluate assertions anymore, because they *may* use a function.

solson replied:
Oh, I didn't read this code carefully enough before.

Should we *ever* visit these? There's no pre-computation to be done for fn items as far as I can tell, so maybe just replace "unimplemented" with a comment about that.

solson commented on src/interpreter/stepper.rs:
```diff
@@ -0,0 +1,214 @@
+            // already computed by rustc
+            mir::Literal::Value { .. } => {}
+            mir::Literal::Item { def_id, substs } => {
+                let item_ty = self.gecx.tcx.lookup_item_type(def_id).subst(self.gecx.tcx, substs);
```
Mmm, I believe `subst` alone is not enough in the case that the item type is an associated type. It should use the `monomorphize` method (on `gecx`, IIRC) which also does associated type normalization.

oli-obk replied:
when I do that, rustc panics about 

> error: internal compiler error: ../src/librustc/infer/mod.rs:650: Encountered errors `[FulfillmentError(Obligation(predicate=Binder(TraitPredicate(<T as std::marker::Sized>)),depth=1),Unimplemented)]`
 resolving bounds after type-checking

solson replied:
Weird. Can you add `// FIXME: Should probably use self.gecx.monomorphize; currently hits an ICE` so I can find it later?

eddyb replied:
Was that from `gecx.monomorphize(constant.ty)`? Cause if it was just the result of `lookup_item_type`, that's not enough.
Generally speaking, `constant.ty == self.gecx.tcx.lookup_item_type(def_id).subst(self.gecx.tcx, substs)` but the latter is more effort.

oli-obk replied:
face -> palm ... I'm fixing this

solson commented on src/interpreter/stepper.rs:
```diff
@@ -163,37 +179,64 @@ impl<'fncx, 'a, 'b: 'a + 'mir, 'mir, 'tcx: 'b> Stepper<'fncx, 'a, 'b, 'mir, 'tcx
+                if self.fncx.frame().promoted.contains_key(&index) {
+                    return;
+                }
+                let mir = self.mir.promoted[index].clone();
```
This is cloning an entire `Mir` structure... we should try to find a way around that.

solson commented on src/interpreter/stepper.rs:
```diff
@@ -0,0 +1,214 @@
+    }
+
+    /// returns the terminator of the current block
+    pub fn term(&self) -> &mir::Terminator {
```
Not a great abbreviation, IMO. s/term/terminator/

oli-obk replied:
I removed the method entirely in a refactoring

solson commented on src/interpreter/mod.rs:
```diff
@@ -118,8 +118,8 @@ enum CachedMir<'mir, 'tcx: 'mir> {
-    /// Make a local jump to the given block.
-    Block(mir::BasicBlock),
+    /// Make a local jump to the next block
+    Block,
```
I don't understand the purpose of this commit.

oli-obk replied:
Before it was that when going to a new block, it was the duty of the `eval_terminator` caller to actually change the block. But if `eval_terminator` did a function call, then `eval_terminator` updated the block (see here: https://github.com/solson/miri/pull/18/files/cc1ca73f5736a4aaa7a1ae164788051b8366fa3d..f995db9ffba925590e4c78917362998be02fcbc1#diff-b6ba58183c703f8ffcaa54879146b1c2R326)

solson replied:
So you were just making `Block` more similar to `Call`? It's not actually required by the rest of the PR?

Just wondering. I don't mind the PR, and I'll probably look for other ways to refactor `TerminatorTarget`.

oli-obk replied:
Well... it caused bugs in the Stepper because I had to very carefully figure out where to get the next block from. Now that I use `Frame::next_block` this might be way less relevant...

I think we might be able to get rid of `TerminatorTarget` completely

solson replied:
> I think we might be able to get rid of TerminatorTarget completely
 
That was my thought. We can merge this PR as-is though.

solson commented on src/interpreter/mod.rs:
```diff
@@ -136,10 +133,28 @@ enum TerminatorTarget {
 
 #[derive(Clone, Debug, Eq, PartialEq, Hash)]
 enum ConstantId<'tcx> {
```
Perhaps make a `struct` that always has `DefId` and `Substs` and avoid the need for the methods.

oli-obk replied:
sounds good

oli-obk replied:
done

solson commented:
Alright, I reviewed nearly all the code. Some of my comments are marked as "on an outdated diff" by GitHub but I think they're all still relevant, aside from one I noticed was already fixed later.

My main areas of concern are the `Stepper` implementation and the new `FrameInterpreter` struct, and I left some comments asking about them.

oli-obk commented:
I addressed all the comments. I modified the commits in place, because some refactorings were easier done by simply removing commits (e.g. the function/frame split)

oli-obk commented:
The major change is the way the stepper works. You pass a closure to the `step` function instead of processing the return value. This makes the code much more concise and prevents us from reloading the entire state at every call. I'm benchmarking currently to figure out if that undoes the performance regression from the constant-jit changes.

oli-obk commented:
how is travis passing if there's a failure? Can we force travis to work with a specific rustc version?

solson commented:
Why do the `Stepper` functions take closures? I don't understand the overall goal with that.

solson commented:
Aside from the closures, `Stepper` makes more sense to me than before.

My main concern now is the `Mir` cloning I mentioned in https://github.com/solson/miri/pull/18#discussion_r66188751.

oli-obk commented:
> What if we inlined the trace! calls into Stepper and only had to worry about done? Could it get simplified down and not require the closures?

done.

> My main concern now is the Mir cloning I mentioned in #18 (comment).

We can cache it to ensure that the cloning only ever happens once for every promoted.

solson commented on src/interpreter/stepper.rs:
```diff
@@ -151,9 +151,10 @@ impl<'a, 'b, 'mir, 'tcx> Visitor<'tcx> for ConstantExtractor<'a, 'b, 'mir, 'tcx>
             // already computed by rustc
             mir::Literal::Value { .. } => {}
             mir::Literal::Item { def_id, substs } => {
-                let item_ty = self.gecx.tcx.lookup_item_type(def_id).subst(self.gecx.tcx, substs);
```
I think this change also applies to https://github.com/solson/miri/pull/18/files#diff-b6ba58183c703f8ffcaa54879146b1c2R1192

oli-obk replied:
it does, fixed

eddyb commented on src/interpreter/stepper.rs:
```diff
@@ -151,9 +151,10 @@ impl<'a, 'b, 'mir, 'tcx> Visitor<'tcx> for ConstantExtractor<'a, 'b, 'mir, 'tcx>
-                let item_ty = self.gecx.tcx.lookup_item_type(def_id).subst(self.gecx.tcx, substs);
-                if item_ty.ty.is_fn() {
-                    // unimplemented
+                if constant.ty.is_fn() {
```
Does this only return true for `TyFnDef`?

oli-obk replied:
good catch, fixed

eddyb commented on src/interpreter/stepper.rs:
```diff
@@ -151,9 +151,10 @@ impl<'a, 'b, 'mir, 'tcx> Visitor<'tcx> for ConstantExtractor<'a, 'b, 'mir, 'tcx>
+                    // because the type is the actual function, not the signature of the function.
+                    // Thus we can simply create a zero sized allocation in `evaluate_operand`
                 } else {
                     self.static_item(def_id, substs, constant.span);
```
Should probably be named `global_item` or some such.

oli-obk replied:
done

solson commented:
I read through the changes one last time. I think I'm ready to merge.

@eddyb Did you have any outstanding comments?

@oli-obk Was there anything you didn't get around to yet that you wanted to?

solson commented:
I heard from eddyb on IRC, so I'll just merge now.
